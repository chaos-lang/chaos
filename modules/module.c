#include "module.h"

void initMainContext() {
    modules_buffer.capacity = 0;
    modules_buffer.size = 0;
    module_path_stack.capacity = 0;
    module_path_stack.size = 0;
    module_stack.capacity = 0;
    module_stack.size = 0;

    char *module_path_with_extension = malloc(1 + strlen(program_file_path));
    strcpy(module_path_with_extension, program_file_path);
    pushModuleStack(module_path_with_extension, "");
    free(module_path_with_extension);
}

void appendModuleToModuleBuffer(char *name) {
    append_to_array(&modules_buffer, name);
    //free(name);
}

void prependModuleToModuleBuffer(char *name) {
    prepend_to_array(&modules_buffer, name);
}

void handleModuleImport(char *module_name, bool directly_import) {
    if (is_interactive) {
        ast_interactive_cursor = ast_node_cursor;
    }
    char *module_path;
    char *relative_path = "";

    module_path = malloc(strlen(module_path_stack.arr[module_path_stack.size - 1]) + 1);
    strcpy(module_path, module_path_stack.arr[module_path_stack.size - 1]);
    if (strchr(module_path, __KAOS_PATH_SEPARATOR_ASCII__) == NULL) {
        free(module_path);
        module_path = "";
    } else {
        char *ptr = strrchr(module_path, __KAOS_PATH_SEPARATOR_ASCII__);
        if (ptr) {
            *ptr = '\0';
        }
    }

    if (module_path[0] != '\0') module_path = strcat_ext(module_path, __KAOS_PATH_SEPARATOR__);

    for (unsigned i = 0; i < modules_buffer.size; i++) {
        module_path = strcat_ext(module_path, modules_buffer.arr[i]);
        relative_path = strcat_ext(relative_path, modules_buffer.arr[i]);
        if (i + 1 != modules_buffer.size) {
            module_path = strcat_ext(module_path, __KAOS_PATH_SEPARATOR__);
            relative_path = strcat_ext(relative_path, __KAOS_PATH_SEPARATOR__);
        }
    }

    char *module = malloc(strlen(modules_buffer.arr[modules_buffer.size - 1]) + 1);
    strcpy(module, modules_buffer.arr[modules_buffer.size - 1]);
    if (module_name != NULL) {
        free(module);
        module = malloc(strlen(module_name) + 1);
        strcpy(module, module_name);
    }
    if (directly_import) {
        free(module);
        module = malloc(strlen("") + 1);
        strcpy(module, "");
    }

    module_path = strcat_ext(module_path, ".");
    module_path = strcat_ext(module_path, __KAOS_LANGUAGE_FILE_EXTENSION__);
    relative_path = strcat_ext(relative_path, ".");
    relative_path = strcat_ext(relative_path, __KAOS_LANGUAGE_FILE_EXTENSION__);

    module_path = searchSpellsIfNotExits(module_path, relative_path);

    pushModuleStack(module_path, module);

    freeModulesBuffer();

    if (strcmp(
        get_filename_ext(module_path),
        __KAOS_DYNAMIC_LIBRARY_EXTENSION__
        ) == 0
    ) {
        callRegisterInDynamicLibrary(module_path);
    } else {
        parseTheModuleContent(module_path);
    }

    freeFunctionNamesBuffer();

    popModuleStack();

    free(module);
    free(module_path);
    free(relative_path);
    //free(module_name);
}

void freeModulesBuffer() {
    for (unsigned i = 0; i < modules_buffer.size; i++) {
        free(modules_buffer.arr[i]);
    }
    if (modules_buffer.size > 0) free(modules_buffer.arr);
    modules_buffer.capacity = 0;
    modules_buffer.size = 0;
}

void pushModuleStack(char *module_path, char *module) {
    append_to_array(&module_path_stack, module_path);
    append_to_array(&module_stack, module);
}

void popModuleStack() {
    free(module_path_stack.arr[module_path_stack.size - 1]);
    module_path_stack.size--;
    free(module_stack.arr[module_stack.size - 1]);
    module_stack.size--;
}

void freeModulePathStack() {
    for (unsigned i = 0; i < module_path_stack.size; i++) {
        free(module_path_stack.arr[i]);
    }
    module_path_stack.capacity = 0;
    module_path_stack.size = 0;
    free(module_path_stack.arr);
}

void freeModuleStack() {
    for (unsigned i = 0; i < module_stack.size; i++) {
        free(module_stack.arr[i]);
    }
    module_stack.capacity = 0;
    module_stack.size = 0;
    free(module_stack.arr);
}

char* getCurrentModule() {
    return module_path_stack.arr[module_path_stack.size - 1];
}

char* getMainModuleDir() {
    char *module_dir;

    module_dir = malloc(strlen(module_path_stack.arr[0]) + 1);
    strcpy(module_dir, module_path_stack.arr[0]);
    if (strchr(module_dir, __KAOS_PATH_SEPARATOR_ASCII__) == NULL) {
        free(module_dir);
        return "";
    }
    char *ptr = strrchr(module_dir, __KAOS_PATH_SEPARATOR_ASCII__);
    if (ptr) {
        *ptr = '\0';
    }

    return module_dir;
}

char* searchSpellsIfNotExits(char* module_path, char* relative_path) {
    module_path = relative_path_to_absolute(module_path);

    if (is_file_exists(module_path)) {
        return module_path;
    } else {
        free(module_path);
        char* spells_dir = getMainModuleDir();
        if (strlen(spells_dir) > 0) spells_dir = strcat_ext(spells_dir, __KAOS_PATH_SEPARATOR__);
        char* spell_name = remove_ext(relative_path, '.', __KAOS_PATH_SEPARATOR_ASCII__);
        spells_dir = strcat_ext(spells_dir, __KAOS_SPELLS__);
        spells_dir = strcat_ext(spells_dir, __KAOS_PATH_SEPARATOR__);
        module_path = strcat_ext(spells_dir, spell_name);
        free(spell_name);
        module_path = strcat_ext(module_path, __KAOS_PATH_SEPARATOR__);
        module_path = strcat_ext(module_path, relative_path);
        module_path = relative_path_to_absolute(module_path);
        if (is_file_exists(module_path)) {
            return module_path;
        } else {
            char* dynamic_library_path = remove_ext(module_path, '.', __KAOS_PATH_SEPARATOR_ASCII__);
            free(module_path);
            dynamic_library_path = strcat_ext(dynamic_library_path, ".");
            dynamic_library_path = strcat_ext(dynamic_library_path, __KAOS_DYNAMIC_LIBRARY_EXTENSION__);
            if (is_file_exists(dynamic_library_path)) {
                return dynamic_library_path;
            } else {
                free(dynamic_library_path);
            }
        }
    }
    printf("HOY\n");
    append_to_array_without_malloc(&free_string_stack, relative_path);
    throw_error(E_MODULE_IS_EMPTY_OR_NOT_EXISTS_ON_PATH, relative_path);
    return NULL;
}
