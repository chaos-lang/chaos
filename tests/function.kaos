void def a()
    print "inside function a"
    num val1 = 5
    print val1
end

print "first print this"

a()
b()

void def b()
    print "inside function b"
    num val2 = 8
    print val2
end

str hello = 'hello world'
num pi = 3.14

str param1 = 'testing'

void def f1(bool param1, num param2, str param3)
    print param1
    print param2
    print param3
end

f1(true, 5, hello)
f1(false, pi, "OMG!")

print hello
print pi

list c = [1, 2, 3]

list def f2(str param1, list param2)
    print param1
    return param2
end

print f2('List function parameter:', c)

dict d = {'a': 'foo', 'b': 'bar', 'c': 'baz'}

dict def f3(str param1, dict param2)
    print param1
    return param2
end

print f3('Dictionary function parameter:', d)

num def add(num x, num y)
    num result = x + y
    return result
end

print add(3, 5)

void def loopInside()
    print "start"
    3 times do
        print "hey"
    end
    print "end"
end

loopInside()

any def anyF1()
    any a = false
    return a
end

print anyF1()

any def anyF2()
    num a = 5
    return a
end

print anyF2()

bool def boolF()
    any a = false
    return a
end

print boolF()

// T_INT inside function
void def tokenIntInsideFunction()
    list a = [1, 2, 3]
    print a
end

tokenIntInsideFunction()

// Tests related to assignments from function returns

import modules.lib


bool def bool_f1()
    bool a = true
    return a
end

bool a = bool_f1()
print a
bool b = lib.bool_f1()
print b


num def num_f1()
    num a = 5
    return a
end

del a
del b
num a = num_f1()
print a
num b = lib.num_f1()
print b


str def str_f1()
    str a = "foo"
    return a
end

del a
del b
str a = str_f1()
print a
str b = lib.str_f1()
print b


any def any_f1()
    any a = 7
    return a
end

del a
del b
any a = any_f1()
print a
any b = lib.any_f1()
print b


list def list_f1()
    list a = [1, 2, 3]
    return a
end

del a
del b
list a = list_f1()
print a
list b = lib.list_f1()
print b


dict def dict_f1()
    dict a = {'a': 1, 'b': 2, 'c': 3}
    return a
end

del a
del b
dict a = dict_f1()
print a
dict b = lib.dict_f1()
print b


// Variable update from function returns

del a
bool a = false
a = bool_f1()
print a
a = lib.bool_f1()
print a

del a
num a = 0
a = num_f1()
print a
a = lib.num_f1()
print a

del a
str a = ''
a = str_f1()
print a
a = lib.str_f1()
print a

del a
any a = 0
a = any_f1()
print a
a = lib.any_f1()
print a

del a
list a = []
a = list_f1()
print a
a = lib.list_f1()
print a

del a
dict a = {}
a = dict_f1()
print a
a = lib.dict_f1()
print a


// Complex element update from function returns

del a
del b
list a = [1, 2, 3, 4, 5, 6]
dict b = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5, 'f': 6}

a[0] = bool_f1()
print a
a[0] = lib.bool_f1()
print a
b['a'] = bool_f1()
print b
b['a'] = lib.bool_f1()
print b

a[1] = num_f1()
print a
a[1] = lib.num_f1()
print a
b['b'] = num_f1()
print b
b['b'] = lib.num_f1()
print b

a[2] = str_f1()
print a
a[2] = lib.str_f1()
print a
b['c'] = str_f1()
print b
b['c'] = lib.str_f1()
print b

a[3] = any_f1()
print a
a[3] = lib.any_f1()
print a
b['d'] = any_f1()
print b
b['d'] = lib.any_f1()
print b

a[4] = list_f1()
print a
a[4] = lib.list_f1()
print a
b['e'] = list_f1()
print b
b['e'] = lib.list_f1()
print b

a[5] = dict_f1()
print a
a[5] = lib.dict_f1()
print a
b['f'] = dict_f1()
print b
b['f'] = lib.dict_f1()
print b


// Direct assignments to typed lists

bool list def bool_list_test()
    bool list x = [true, false]
    return x
end
bool list bool_list_test_var = bool_list_test()
print bool_list_test_var

num list def num_list_test()
    num list x = [1, 3.14]
    return x
end
num list num_list_test_var = num_list_test()
print num_list_test_var

str list def str_list_test()
    str list x = ['a', "b"]
    return x
end
str list str_list_test_var = str_list_test()
print str_list_test_var

// Direct assignments to typed dictionaries

bool dict def bool_dict_test()
    bool dict x = {'a': true, 'b': false}
    return x
end
bool dict bool_dict_test_var = bool_dict_test()
print bool_dict_test_var

num dict def num_dict_test()
    num dict x = {'a': 1, 'b': 3.14}
    return x
end
num dict num_dict_test_var = num_dict_test()
print num_dict_test_var

str dict def str_dict_test()
    str dict x = {'a': 'a', "b": "b"}
    return x
end
str dict str_dict_test_var = str_dict_test()
print str_dict_test_var


// Optional function parameters

bool def bool_optional_param_test(bool param1, bool param2 = true)
    print param1
    print param2
    return param1
end

bool_optional_param_test(true)
bool_optional_param_test(false)

bool_optional_param_test(true, false)
bool_optional_param_test(false, true)

num def num_optional_param_test(num param1, num param2 = 7)
    print param1
    print param2
    return param1
end

num_optional_param_test(3)
num_optional_param_test(3.14)
num_optional_param_test(3, 5)
num_optional_param_test(3, 5.12)
num_optional_param_test(3.14, 5.12)

str def str_optional_param_test(str param1, str param2 = 'baz')
    print param1
    print param2
    return param2
end

print str_optional_param_test('foo')
print str_optional_param_test('foo', 'bar')


// Lists and dictionaries as function parameters

void def list_as_params(str param1, list param2, list param3 = [4, 5, 6])
    print param1
    print param2
    print param3
end

list_as_params('foo', [1, 2, 3])
list_as_params('bar', [1, 2, 3], [7, 8, 9])

void def dict_as_params(str param1, dict param2, dict param3 = {'d': 4, 'e': 5, 'f': 6})
    print param1
    print param2
    print param3
end

dict_as_params('foo', {'a': 1, 'b': 2, 'c': 3})
dict_as_params('bar', {'a': 1, 'b': 2, 'c': 3}, {'j': 7, 'k': 8, 'l': 9})


// Typed lists and typed dictionaries as function parameters

void def typed_list_as_params(str param1, num list param2, num list param3 = [4, 5, 6])
    print param1
    print param2
    print param3
end

typed_list_as_params('foo', [1, 2, 3])
typed_list_as_params('bar', [1, 2, 3], [7, 8, 9])

void def typed_dict_as_params(str param1, num dict param2, num dict param3 = {'d': 4, 'e': 5, 'f': 6})
    print param1
    print param2
    print param3
end

typed_dict_as_params('foo', {'a': 1, 'b': 2, 'c': 3})
typed_dict_as_params('bar', {'a': 1, 'b': 2, 'c': 3}, {'j': 7, 'k': 8, 'l': 9})


// Call function inside loop inside function

num def call_inside_loop_inside_function(num n)
	print n
	num p = n+1
	return p
end

void def test_call_inside_loop_inside_function()
	num x = 0
    num r = 0
	3 times do
		r = call_inside_loop_inside_function(x)
        print r
	end
end

test_call_inside_loop_inside_function()

// Call parameters in decision block

num def multi_three(num x, num y, num z)
    num r = x * y * z
    return r
end

num def sum_two(num x, num y)
    num r = x + y
    return r
end

num def decision_params(num n)
    num a = 2
    num b = 3
    num c = 5
end {
    n == 0  : sum_two(a, b),
    n == 1  : sum_two(b, c),
    default : multi_three(a, b, c)
}

print decision_params(0)
print decision_params(1)
print decision_params(2)
