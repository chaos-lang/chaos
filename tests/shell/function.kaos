import tests.shell.modules.module1

void def fa()
    print "hey"
end

asdasd1()

fa()

asdasd2()

num def fb()
    num a = 5
    return a
end
print fb()

// Print the function return on function calls from modules
print module1.sub(7, 3)

// Test for not printing function returns on function decision execution mode
num def f1()
    num a = 101
    return a
end

num def f2()
    num b = 102
    return b
end

num def f3()
    num c = 103
    return c
end

num def add(num x, num y)
    num z = x + y
end {
    z == 8  : f1(),
    z > 10  : f2(),
    default : f3()
}

print add(3, 5)
