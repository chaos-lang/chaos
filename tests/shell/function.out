Chaos Error:
Module: ~/chaos/__interactive__.kaos
Line: 15
Undefined function: asdasd1
Absorbed by Interactive Shell
hey
Chaos Error:
Module: ~/chaos/__interactive__.kaos
Line: 4
Undefined function: asdasd2
Absorbed by Interactive Shell
5
4
101
101
